packages/eslint-plugin/src/rules/no-empty-object-type.ts

```diff
@@ -0,0 +1,166 @@
+
+import { createRule } from '../util';
+
+export type AllowInterfaces = 'always' | 'never' | 'with-single-extends';
+
+export type AllowObjectTypes = 'always' | 'never';
+
+export type Options = [
+  {
+    allowInterfaces?: AllowInterfaces;
+    allowObjectTypes?: AllowObjectTypes;
+  },
+];
+
+export type MessageIds =
+  | 'noEmpty'
+  | 'noEmptyInterfaceWithSuper'
+  | 'replaceEmptyInterface'
+  | 'replaceEmptyInterfaceWithSuper'
+  | 'replaceEmptyObjectType';
+
+export default createRule<Options, MessageIds>({
+  name: 'no-empty-object-type',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow accidentally using the "empty object" type',
+      recommended: 'recommended',
+    },
+    hasSuggestions: true,
+    messages: {
+      noEmpty: [
+        'The `{}` ("empty object") type allows any non-nullish value, including literals like `0` and `""`.',
+        "- If that's what you want, disable this lint rule with an inline comment or configure the '{{ option }}' rule option.",
+        '- If you want a type meaning "any object", you probably want `object` instead.',
+        '- If you want a type meaning "any value", you probably want `unknown` instead.',
+      ].join('\n'),
+      noEmptyInterfaceWithSuper:
+        'An interface declaring no members is equivalent to its supertype.',
+      replaceEmptyInterface: 'Replace empty interface with `{{replacement}}`.',
+      replaceEmptyInterfaceWithSuper:
+        'Replace empty interface with a type alias.',
+      replaceEmptyObjectType: 'Replace `{}` with `{{replacement}}`.',
+    },
+    schema: [
+      {
+        type: 'object',
+        additionalProperties: false,
+        properties: {
+          allowInterfaces: {
+            enum: ['always', 'never', 'with-single-extends'],
+            type: 'string',
+          },
+          allowObjectTypes: {
+            enum: ['always', 'never'],
+            type: 'string',
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      allowInterfaces: 'never',
+      allowObjectTypes: 'never',
+    },
+  ],
+  create(context, [{ allowInterfaces, allowObjectTypes }]) {
+    return {
+      ...(allowInterfaces !== 'always' && {
+        TSInterfaceDeclaration(node): void {
+          const extend = node.extends;
+          if (
+            node.body.body.length !== 0 ||
+            (extend.length === 1 &&
+              allowInterfaces === 'with-single-extends') ||
+            extend.length > 1
+          ) {
+            return;
+          }
+
+          const scope = context.sourceCode.getScope(node);
+
+          const mergedWithClassDeclaration = scope.set
+            .get(node.id.name)
+            ?.defs.some(
+              def => def.node.type === AST_NODE_TYPES.ClassDeclaration,
+            );
+
+          if (extend.length === 0) {
+            context.report({
+              data: { option: 'allowInterfaces' },
+              node: node.id,
+              messageId: 'noEmpty',
```

Comment:
do we want a different message for interfaces?
otherwise the report can be a little confusing:
```ts
interface Foo {}
^^^^^^^^^^^^^^^^ The `{}` ("empty object") type allows any non-nullish value, including literals like `0` and `""`.
```

"But I'm not using the empty object type - I'm declaring an interface! What gives?!?"

Might be good to parameterise this for the two usecases eg
``The `{}` ("empty object") type ...``
and
``An empty interface declaration ...``